File: back/ScadaSnusProject/Services/TagService.cs
using System.Collections;
using ScadaSnusProject.Model;
using ScadaSnusProject.Repositories.Interfaces;
using ScadaSnusProject.Services.Interfaces;

namespace ScadaSnusProject.Services;

public class TagService : ITagService
{
    private readonly ITagRepository _tagRepository;

    public TagService(ITagRepository tagRepository)
    {
        _tagRepository = tagRepository;
    }
    public void AddAnalogInput(AnalogInput analogInput)
    {
        var success = _tagRepository.AddAnalogInput(analogInput);
        if (!success)
        {
            throw new Exception("This IOAddress is already in use!");
        }
    }

    public void AddAnalogOutput(AnalogOutput analogOutput)
    {
        var success = _tagRepository.AddAnalogOutput(analogOutput);
        if (!success)
        {
            throw new Exception("This IOAddress is already in use!");
        }
    }

    public void AddDigitalInput(DigitalInput digitalInput)
    {
        var success = _tagRepository.AddDigitalInput(digitalInput);
        if (!success)
        {
            throw new Exception("This IOAddress is already in use!");
        }
    }

    public void AddDigitalOutput(DigitalOutput digitalOutput)
    {
        var success = _tagRepository.AddDigitalOutput(digitalOutput);
        if (!success)
        {
            throw new Exception("This IOAddress is already in use!");
        }
    }

    public void DeleteTag(int tagId)
    {
        var success = _tagRepository.DeleteTag(tagId);
        if (!success)
        {
            throw new Exception("Tag with this id does not exist!");
        }
    }

    public ICollection<Tag> GetAllInputs()
    {
        return _tagRepository.GetAllInputs();
    }

    public ICollection<Tag> GetAllOutputs()
    {
        return _tagRepository.GetAllOutputs();
    }

    public void TurnOnScan(int tagId)
    {
        var success = _tagRepository.TurnOnScan(tagId);
        if (!success)
        {
            throw new Exception("Tag with this id does not exist or is already turned on!");
        }
    }

    public void TurnOffScan(int tagId)
    {
        var success = _tagRepository.TurnOffScan(tagId);
        if (!success)
        {
            throw new Exception("Tag with this id does not exist or is already turned off!");
        }
    }

    public ICollection<Tag> GetAllTags()
    {
        return _tagRepository.GetAllTags();
    }

    public ICollection<DigitalInput> GetAllDigitalInputs()
    {
        return _tagRepository.GetAllDigitalInputs();
    }

    public ICollection<DigitalOutput> GetAllDigitalOutputs()
    {
        return _tagRepository.GetAllDigitalOutputs();
    }

    public ICollection<AnalogInput> GetAllAnalogInputs()
    {
        return _tagRepository.GetAllAnalogInputs();
    }

    public ICollection<AnalogOutput> GetAllAnalogOutputs()
    {
        return _tagRepository.GetAllAnalogOutputs();
    }

    public ICollection<Tag> GetAllOnScanInputs()
    {
        return _tagRepository.GetAllOnScanInputs();
    }
}

File: front/SCADA_project/src/app/services/tag/tag.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { HttpTransportType, HubConnection, HubConnectionBuilder } from '@microsoft/signalr';
import { Observable } from 'rxjs';
import { environment } from 'src/enviroments/environment';

@Injectable({
  providedIn: 'root'
})
export class TagService {
  private hubConnection: HubConnection;

  constructor(private http: HttpClient) { 
    this.hubConnection = new HubConnectionBuilder()
      .withUrl("https://localhost:7031/tagValuesHub", {skipNegotiation:true, transport: HttpTransportType.WebSockets})
      .build();
  }

  startConnection() {
    return this.hubConnection.start();
  }

  stopConnection() {
    return this.hubConnection.stop();
  }

  getHubConnection() {
    return this.hubConnection;
  }

  addAnalogInput(aInput: AnalogInput): Observable<any> {
    return this.http.post<any>(environment.apiHost + 'tag/add-analog-input', aInput);
  }

  addAnalogOutput(aOutput: AnalogOutput): Observable<any> {
    return this.http.post<any>(environment.apiHost + 'tag/add-analog-output', aOutput);
  }

  addDigitalInput(dInput: DigitalInput): Observable<any> {
    return this.http.post<any>(environment.apiHost + 'tag/add-digital-input', dInput);
  }

  addDigitalOutput(dOutput: DigitalOutput): Observable<any> {
    return this.http.post<any>(environment.apiHost + 'tag/add-digital-output', dOutput);
  }

  removeTag(tagId: number): Observable<any> {
    return this.http.delete<any>(environment.apiHost + 'tag/remove-tag-' + tagId);
  }

  turnOnScan(tagId: number): Observable<any> {
    return this.http.put<any>(environment.apiHost + 'tag/scan-on-' + tagId, {});
  }

  turnOffScan(tagId: number): Observable<any> {
    return this.http.put<any>(environment.apiHost + 'tag/scan-off-' + tagId, {});
  }

  getAllTags(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag');
  }

  getAllInputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-inputs');
  }

  getAllOutputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-outputs');
  }

  getAllAnalogInputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-analog-inputs');
  }

  getAllAnalogOutputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-analog-outputs');
  }

  getAllDigitalInputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-digital-inputs');
  }

  getAllDigitalOutputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-digital-outputs');
  }

  getAllOnScanInputs(): Observable<any> {
    return this.http.get<any>(environment.apiHost + 'tag/all-on-scan-inputs');
  }
}

export interface AnalogInput{
  id?: number,
  name: string,
  description: string,
  ioAddress: string,
  value: number,
  scanTime: number,
  isScanOn: boolean,
  lowLimit: number,
  highLimit: number,
  unit: string
}

export interface AnalogOutput{
  id?: number,
  name: string,
  description: string,
  ioAddress: string,
  value: number,
  lowLimit: number,
  highLimit: number
  unit: string
}

export interface DigitalInput{
  id?: number,
  name: string,
  description: string,
  ioAddress: string,
  value: number,
  scanTime: number,
  isScanOn: boolean
}

export interface DigitalOutput{
  id?: number,
  name: string,
  description: string,
  ioAddress: string,
  value: number
}

export interface Tag{
  id: number,
  name: string,
  description: string,
  ioAddress: string,
  value: number
}

export interface TagValue{
  id: number,
  timestamp: string,
  value: number,
  tagId: number
}